### PR TITLE
Add cognitive complexity metric (TS) and rule (TS/JS) based on 'cognitive-complexity' from eslint-plugin-sonarjs

### LINKED ISSUE #1081: Make CognitiveComplexity(S3776) eslint-based rule (cognitive-complexity) (TS/JS)
Currently we don't have a way to support secondary issue locations for eslint-based rules.
Since those are fundamental for cognitive complexity, we will replace this rule when we will implement secondary locations for eslint-based rules.

### LINKED ISSUE #1326: Provide cognitive complexity metrics for TypeScript

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes #1326 
Fixes #1081

### CODE REVIEW

File: eslint-bridge/src/linter.ts [hunk 1]
```diff
@@ -49,14 +55,16 @@ export function analyze(
   sourceCode: SourceCode,
   inputRules: Rule[],
   fileUri: string,
-  additionalRule?: { ruleId: string; ruleModule: ESLintRule.RuleModule },
+  ...additionalRules: AdditionalRule[]
 ) {
-  if (additionalRule) {
-    linter.defineRule(additionalRule.ruleId, additionalRule.ruleModule);
+  if (additionalRules.length > 0) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/linter.ts]: We can remove the `if` statement.

File: eslint-bridge/src/linter.ts [hunk 2]
```diff
@@ -110,16 +118,19 @@ function removeIrrelevantProperties(eslintIssue: Linter.LintMessage): Issue | nu
...
 ) {
   const ruleConfig: Linter.Config = { rules: {}, parserOptions: { sourceType: "module" } };
   inputRules.forEach(inputRule => {
     const ruleModule = linter.getRules().get(inputRule.key);
     ruleConfig.rules![inputRule.key] = ["error", ...getRuleConfig(ruleModule, inputRule)];
   });
 
-  if (additionalRule) {
-    ruleConfig.rules![additionalRule.ruleId] = ["error"];
+  if (additionalRules.length > 0) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/linter.ts]: We can remove the `if` statement here too.

File: eslint-bridge/src/analyzer.ts [hunk 3]
```diff
@@ -123,15 +124,54 @@ function analyze(input: AnalysisInput, parse: Parse): AnalysisResponse {
...
+    input.filePath,
+    SYMBOL_HIGHLIGHTING_RULE,
+    COGNITIVE_COMPLEXITY_RULE,
+  ).issues;
+  const highlightedSymbols = getHighlightedSymbols(issues);
+  const metrics = getMetrics(sourceCode);
+  metrics.cognitiveComplexity = getCognitiveComplexity(issues);
+  return {
+    issues,
+    highlights: getHighlighting(sourceCode).highlights,
```
Comment (saberduck) [inline eslint-bridge/src/analyzer.ts]: wdyt about extracting all properties to variables in this object literal?

File: eslint-bridge/src/analyzer.ts [hunk 4]
```diff
@@ -123,15 +124,54 @@ function analyze(input: AnalysisInput, parse: Parse): AnalysisResponse {
...
+  } else {
+    console.log("DEBUG Failed to retrieve symbol highlighting from analysis results");
+    return [];
+  }
+}
+
+// exported for testing
+export function getCognitiveComplexity(issues: Issue[]) {
+  const issue = findAndRemoveFirstIssue(issues, COGNITIVE_COMPLEXITY_RULE_ID);
+  if (issue && !isNaN(Number(issue.message))) {
```
Comment (saberduck) [inline eslint-bridge/src/analyzer.ts]: can we use `Number.isNan` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN ?

File: eslint-bridge/src/linter.ts [hunk 5]
```diff
@@ -49,14 +55,16 @@ export function analyze(
   sourceCode: SourceCode,
   inputRules: Rule[],
   fileUri: string,
-  additionalRule?: { ruleId: string; ruleModule: ESLintRule.RuleModule },
+  ...additionalRules: AdditionalRule[]
```
Comment (saberduck) [inline eslint-bridge/src/linter.ts]: some comment about purpose of this extra parameter could be helpfull

File: eslint-bridge/src/analyzer.ts (hunk 4 above)
Comment (vilchik-elena) [reply eslint-bridge/src/analyzer.ts]: In fact I believe we should use just `isNaN` (`Number.isNan` results false if argument is not number)

Comment (vilchik-elena) [reply eslint-bridge/src/analyzer.ts]: in fact both these function in TS require number as argument, so it's ok this way

**Pull request merged**